### PR TITLE
Participant pruning ledger api server support ported from Canton

### PR DESCRIPTION
**Pull request by oliverse-da**
Please refer to the participant pruning design document as you review this, @rautenrieth-da .

Also pending: A `ParticipantPruningIT` test suite similar to the one in the predecessor PR to this one: https://github.com/digital-asset/daml/pull/5802

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala
Comment (stefanobaghino-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala:60]: This doesn't look like it's adding anything. Why not calling the method above directly?

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres/V39__Participant-pruning.sql
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/postgres/V39__Participant-pruning.sql:12]: If `NULL` means "no prune happened so far" this should be fine. Is that so?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:43]: Please coordinate with @SamirTalwar-DA on [this PR](https://github.com/digital-asset/daml/pull/7945).

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:50]: Should this be wrapped in a failed future instead of just throwing?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:77]: A few things here:

1. break down each step into a method
2. move further validation down below here
3. refactor the logic into its own method (could be named `validateRequest` or something like this)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:88]: As mentioned before, keep the validation logic together in the block above.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:102]: This is also a good candidate to be put into its own method.

Ideally you would end up with something like:

```scala
val pruneUpTo =
  for {
    _ <- checkPruneUpTo(request)
    pruneUpTo <- decodePruneUpTo(request)
    _ <- validatePruneUpTo(pruneUpTo)
  } yield prunedUpTo

prunedUpTo.fold(Future.failed, doPrune(writeBackend, readBackend, submissionId, _))
```

Or something like this.

File: ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/LedgerBackedWriteService.scala
Comment (stefanobaghino-da) [inline ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/LedgerBackedWriteService.scala:104]: I'm not sure "sandbox classic" is a thing anymore.

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres/V39__Participant-pruning.sql
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres/V39__Participant-pruning.sql:12]: Correct

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:43]: Yeah, I have been monitoring that PR.

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:43]: As this one is likely to make it in first, I suggest you just comment on the other PR once it's merged so I don't forget to update this. (Or push an update there, if you like.)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala]: Okay, I took a stab at putting each of the steps (and to the note above: `validateRequest` as a wrapper function for validation) after trying out a few arrangements, I think I ended up with the simplest, most readable one by keeping the submissionId at the outermost level, so that I only end up with one `LoggingContext` wrapper.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala]: I followed up on this and added a note below. Resolving this and keeping the note below open for tracking.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:135]: Could be factored out as a separate value, next to the related `offsetToPruneUpTo` and `offsetOfFirstSurvivingTransaction` values. Maybe as a `offsetOfLastRemovedTransaction`?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:323]: Can this ever fail? `prunedTransactionIds` is constructed from `Range(itemsToPrune - transactionsPerBatch, itemsToPrune)`, which has always size `transactionsPerBatch`.

The real check seems to be that the corresponding `Future.sequence(ids.map(participant.transactionTreeById))` succeeds.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:525]:
```suggestion
          _ <- participant.prune(offsetToPruneUpToInSecondRealPrune)
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:530]:
```suggestion
              .update(_.begin := offsetOfFirstSurvivingTransactionInSecondPrune))
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:63]: If the write service pruning succeeds but the index pruning fails, then the participant is probably left in a corrupt state.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:85]: As a failed admin operation, it should probably print a log line at the highest log level.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:94]: Same here, print a log line if the pruning fails.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala
Comment (rautenrieth-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:21]: > return the input offset as the output

`ParticipantPruned` does not contain any offset to be returned, I assume this is a leftover from a previous iteration?

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala
Comment (rautenrieth-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:18]: As noted above, we should probably talk more about failed pruning operations. Maybe also advise WriteService implementations to perform any destructive operations atomically (e.g., in one SQL transaction).

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PruningResult.scala
Comment (rautenrieth-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PruningResult.scala:12]: > Indicates readiness to prune the ledger api server

What happens if the ledger API server does not prune its index database at the same offset?

**Review by rautenrieth-da — COMMENTED** [review]
Looks good, left a few comments.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:669]: I will add the following comment:

```
  def prune(pruneUpTo: Option[LedgerOffset], attempts: Int): Future[Empty] =
+    // Distributed ledger participants need to reach global consensus prior to pruning. Hence the "eventually" here:
    eventually(attempts = attempts, runAssertion = {
      services.participantPruning.prune(PruneRequest(pruneUpTo, nextSubmissionId()))
    })
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:111]: It looks like `ParticipantTestContext.getTransactionsRequest` sets the end to be `Some(LedgerBoundary.LEDGER_END)` and internally to the ledger api server, `LedgerBackedIndexService.convertOffset` converts this to the actual ledger end.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:323]: You are right that the two size-based asserts are tautologies and as such may be confusing. My intent was to express that the last `n` pruned transactions are now invisible and the subsequent `n` unpruned transactions are still visible. I can get rid of these asserts.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:530]: Good catch above - I missed that. This instance however should be `offsetToPruneUpToInSecondRealPrune` as begin is an exclusive start.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:85]: Agreed with logging an error in these cases. This should be taken care of by this expression following the future-for-comprehension in the caller or?: `.andThen(logger.logErrorsOnCall[Empty])`

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:21]: It is - thanks. I must have removed this from at least a dozen of places. Thank you for the careful review.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:94]: See above.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:18]: I have written up a section on fault tolerance taking your points as inspiration. Atomic-SQL transactions are one model and another one is breaking up pruning into idempotent steps in such a way that they lend themselves to reaching eventual consistency upon failures.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PruningResult.scala
Comment (oliverse-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PruningResult.scala:12]: Let's sync and see if the blurb I am adding to the interface addresses your question or if there is another nuance I may be missing.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (rautenrieth-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala:111]: Ah, that's the part I missed, thanks!

**Review by gerolf-da — COMMENTED** [review]
First batch of comments :)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (rautenrieth-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:87]: Ah right, I missed that part. There is one small detail, `logger.logErrorsOnCall` may not contain details on which part of the pruning failed (e.g., whether it was the index pruning or the write service pruning), but I trust WriteService implementations to put enough details in the Status message.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PruningResult.scala
Comment (rautenrieth-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PruningResult.scala:14]: Purely cosmetic: You could wrap the two cases in a `object PruningResult`, to allow users to write `PruningResult. ParticipantPruned`. We use the same pattern for [state updates](https://github.com/digital-asset/daml/blob/master/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala#L29)

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (rautenrieth-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:670]: Discussed online: some ledgers need to perform ledger-wide synchronization before allowing participants to delete local data. I would suggest one of the following:

- Require `WriteParticipantPruningService.prune` implementations to wait a resonable time until they are ready to prune. This would remove the need for `eventually`. On the other hand, the ledger API call could take a long time and possibly trigger some network timeout.
- Clearly document in `participant_pruning_service.proto` and in `PruningResult.scala` which error code means that the participant is not ready yet and that the client application should retry later at the same offset. You could even consider adding a `PruningResult.WorkingOnItPleaseRetryLater` case for this.

Speaking of which, the protobuf file could use comments on what error messages the prune call can return, see the [command service example](https://github.com/digital-asset/daml/blob/master/ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto#L24).

**Review by gerolf-da — COMMENTED** [review]
Second review batch. Thanks, @oliverse-da!

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:60]: Ok, I will add it inside the helper functions `pruneWriteService` and `pruneLedgerApiServerIndex` where I have already been logging completion of these steps, but not the beginning.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/QueryNonPruned.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/QueryNonPruned.scala:13]: In the four streaming queries we want to produce an error rather than filtering the results, as that is the mechanism by which we make users aware of a potentially broken app (that is either unaware of pruning and the resulting inability to read the ledger from `begin` or has fallen behind the pruning schedule).

However in the two point-id queries, we do filter, and those cases we could handle by doing the join in the database.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/QueryNonPruned.scala
Comment (oliverse-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/QueryNonPruned.scala:46]: How about `executeSqlOrThrow`?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala
Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala:46]: Thanks for the pointer. At some point I had been concerned that I would break other test suites particularly ones that read from the beginning of the ledger. I only found two such test (`TransactionServiceIT` "TXBeginToBegin" and "TXTreesBeginToBegin") cases after looking, but they only use empty ranges which do not result in pruning-related errors, so using this flag should work.

As a second layer of defense (that I don't think we want to rely on), the sequential tests are run after the concurrent tests, so if they leave a participant in such a state that concurrent tests no longer pass, that would be hidden from us. For now, I am adding a somewhat artificial test `ParticipantPruningIT."PRPruneQueryEmptyRangeOk"` that notifies us if querying empty ranges in pruned offsets start to error.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:670]: Good feedback, I am adding a section on error codes to both the `participant_pruning_service.proto` (guidance to users what errors to expect) and to `PruningResult.scala` (guidance to pruning implementers on what types of errors to emit).

Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:670]: > Would a new request trigger a new global consensus process, or can it use the result of a previews round?

Consensus gathering happens in the background async to  prune requests. The `eventually` with subsequent `prune` requests simply wait for the consensus to proceed beyond the specific offset in the background.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/QueryNonPruned.scala
Comment (gerolf-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/QueryNonPruned.scala:13]: I see. thanks.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto
Comment (oliverse-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto]: Changed.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala
Comment (oliverse-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala]: Change accordingly.

File: compatibility/bazel_tools/testing.bzl
Comment (gerolf-da) [inline compatibility/bazel_tools/testing.bzl:156]: I assume this still needs to be changed before the merge?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:95]:
```suggestion
    logger.info(s"About to prune ledger api server index to ${pruneUpTo} inclusively")
```

Comment (gerolf-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:95]: Please `import ApiOffset.ApiOffsetSConverter` and use `${pruneUpTo.toApiString}` in the log messages to make the offset somewhat readable.

File: compatibility/bazel_tools/testing.bzl
Comment (oliverse-da) [reply compatibility/bazel_tools/testing.bzl:156]: The way I understand the protocol for compatibility testing is that  these values will be filled in as part of the release. I.e. this value would become the first release that includes the `ParticipantPruningIT` test suite. Is that right @moritzkiefer-da ?

Comment (cocreature) [reply compatibility/bazel_tools/testing.bzl:156]: Yeah it’s a bit annoying, we should add exclusive bounds but atm you’re stuck with merging this and then fixing it up on the next snapshot.

**Review by gerolf-da — APPROVED** [review]
Thank you very much, @oliverse-da!

**Review by rautenrieth-da — APPROVED** [review]

**Pull request merged**
